R/grade_code.R

```diff
@@ -79,13 +79,19 @@ grade_code <- function(
   glue_incorrect = getOption("gradethis_glue_incorrect"),
   glue_pipe = getOption("gradethis_glue_pipe")
 ) {
-  user <- grader_args$user_quo
-  solution <- grader_args$solution_quo
+  user <- rlang::as_quosure(grader_args$user_quo)
+  solution <- rlang::as_quosure(grader_args$solution_quo)
 
-  # MUST call user first to avoid "poisoning" the envir with solution information
   user <- rlang::get_expr(user)
   solution <- rlang::get_expr(solution)
 
+  if (!is.null(user)) {
+    stopifnot(is.expression(user))
+  }
+  if (!is.null(solution)) {
+    stopifnot(is.expression(solution))
+  }
+
   if (is_code_identical(user, solution)) {
     is_same_info <- graded(correct = TRUE, message = NULL)
   } else {
```

Comment:
Change `detect_mistakes(user, solution)` to `detect_mistakes(user, solution, env = learnr_args$envir_result)`

Reply:
Maybe add a helper since we do this in multiple places

```r
learnr_env <- function(args = NULL) {
  args$envir_result %||% args$envir_prep %||% stop("Internal error. learnr did not pass a relevant environment")
}
```